neptune/alpha/internal/backends/operations_preprocessor.py

```diff
@@ -0,0 +1,190 @@
+#
+# Copyright (c) 2020, Neptune Labs Sp. z o.o.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from enum import Enum
+from typing import List, TypeVar, Callable
+
+from neptune.alpha.exceptions import MetadataInconsistency, InternalClientError
+from neptune.alpha.internal.operation import Operation, AssignFloat, AssignString, UploadFile, LogFloats, LogStrings, \
+    LogImages, ClearFloatLog, ClearStringLog, ClearImageLog, AddStrings, RemoveStrings, DeleteAttribute, ClearStringSet
+from neptune.alpha.internal.operation_visitor import OperationVisitor
+from neptune.alpha.internal.utils.paths import path_to_str
+
+T = TypeVar("T")
+
+
+class OperationsPreprocessor:
+
+    def __init__(self):
+        self._accumulators = dict()
+
+    def process(self, operations: List[Operation]) -> List[Operation]:
+        for op in operations:
+            path_str = path_to_str(op.path)
+            self._accumulators.get(path_str, _OperationsAccumulator(op.path)).visit(op)
+        result = []
+        for acc in self._accumulators.values():
+            result.extend(acc.get_operations())
+        return result
+
+
+class _DataType(Enum):
+    FLOAT = "Float",
+    STRING = "String",
+    FILE = "File"
+    FLOAT_SERIES = "Float Series",
+    STRING_SERIES = "String Series",
+    IMAGE_SERIES = "Image Series",
+    STRING_SET = "String Set"
+
+
+class _OperationsAccumulator(OperationVisitor[None]):
```

Comment:
This algorithm is nice and generic (I'll certainly borrow some ideas for the Java version) but I can see some potential problems: there is a property which I'll sketch below which an accumulating/compressing algorithm should respect, and I'm not such this algorithm respects this property.

Namely, compressed operations should be equivalent to the original operations, regardless of the state of the attribute they act on. But the client does not know the experiment's state on the server, so it needs to compress in a way which works for all possible server states.

For example, suppose that an experiment has a `FloatSeries([1.0, 1.1])` attribute defined at path `path`.

Now suppose that we have an uncompressed list of operations `[AssignFloat(path, 1.2), LogFloats(path, [1.3, 1.4])]`.

If you send the original list to the server, then the server will reject the `AssignFloat` operation with an error, and process the `LogFloats` operation ending up in a state with `path` set to `FloatSeries([1.0, 1.1, 1.3, 1.4])`.

But your aggregation algorithm will compress those original operations to just `[AssignFloat(path, 1.2)]`, rejecting the `LogFloats` operation.

If you send `[AssignFloat(path, 1.2)]` to the server, then the server will reject that operation with an error and end up in a state with `path` set to `FloatSeries([1.0, 1.1])`, which is a different state.

An example of a client-side aggregation which works for all possible server states is compressing

`[AssignFloat(path, 1.2), LogFloats(path, [1.3, 1.4]), LogFloats(path, [1.5, 1.6])]`

to 

`[AssignFloat(path, 1.2), LogFloats(path, [1.3, 1.4, 1.5, 1.6])]`